### ISSUE TITLE
hints: Handle hints close to each other

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#71]
- [ ] When there are two links to the same location and the hints are very close or even overlapping (e.g. icon/text), one should be hidden. (#1955 also proposes merging them for auto-following when filtering by text)

Note from @flvi0:

> I'm not a web developer but I just reassured with my brother who is, a page can still detect which of the links was clicked using javascript and potentially behave differently even if they point to the same url, so merging them seems error-prone to me.
- [ ] When there are two overlapping hints with different links, maybe they should get a slight X/Y offset or so? See youtube videos on facebook for example: 

![youtube](https://cloud.githubusercontent.com/assets/625793/6186244/442e9884-b377-11e4-90c4-e0daca023871.png)
- [ ] For multiple hints with the same link (but not close), they could share the same keybinding (PR #5287)

**Comment by Yousuf-Khurshid** [comment]
I know this isn't exactly hig priority atm but could it be possible to simply add an optional keybinding (such as perhaps Shift) to just flip the order of overlapping hints. I believe this same solution was used in the Chrome extension SurfingKeys.

**Comment by The-Compiler** [comment]
@joemaro No, but there's always a lot more to do than there's time - so as long as nobody picks this up, nothing will happen :wink:

**Comment by tobip** [comment]
> a page can still detect which of the links was clicked using javascript and potentially behave differently even if they point to the same url, so merging them seems error-prone to me.

Even if that is the case I would love to have, for example, an option in the qutebrowser settings to turn on/off giving same hints to same urls. 

Because for users like me the benefit outweighs the risk of having an unexpected behavior on some sites.

As a sidenote: In my opinion a webpage does not follow a good UX/Design when it contains the same url multiple times. I can imagine that screen reading software used by eg. blind people struggles with similar difficulties.

**Comment by The-Compiler** [comment]
FWIW, for some common cases (link targets like `#`, `javascript:`, etc.) I guess we could turn off the deduplication.

**Comment by bovine3dom** [comment]
At the same time we added https://github.com/tridactyl/tridactyl/pull/2534 so that users can tell which hints correspond to JS (and are therefore generally useless). The overlap had previously done that automatically as useless JS hints were often covered up by real hints.

Disclaimer: that PR added some subtle bugs to our hint mode that I haven't fixed yet.

**Comment by pacien** [comment]
Here are two instances of this issue that I'm encountering daily:

- Wikipedia articles, in which references `[1][2][3]...` are very close to
  each other.

  ![Kooha-2023-05-29-05-59-00](https://github.com/qutebrowser/qutebrowser/assets/1449319/16f69970-c8a1-458f-a942-426fe347a335)


- Haskell documentation anchor and source links:
  even more problematic, it's not possible to select the hints for the "Source"
  links because the beginning is hidden by the overlapping anchor hints.
  [Example page](https://hackage.haskell.org/package/checkers-0.6.0/docs/Test-QuickCheck-Classes.html).

  ![Kooha-2023-05-29-06-00-00](https://github.com/qutebrowser/qutebrowser/assets/1449319/b348a677-1ebd-4b42-81bd-124c106c9e31)
